--- libiwinfo-2018-07-24-94b1366d/iwinfo_nl80211.c	2018-07-24 21:06:22.000000000 +0530
+++ new-libiwinfo-2018-07-24-94b1366d/iwinfo_nl80211.c	2019-03-13 21:38:56.982957406 +0530
@@ -864,7 +864,8 @@ static int __nl80211_wpactl_query(const
 	if (nl80211_get_mode(ifname, &mode))
 		return 0;
 
-	if (mode != IWINFO_OPMODE_CLIENT && mode != IWINFO_OPMODE_ADHOC)
+	if (mode != IWINFO_OPMODE_CLIENT && mode != IWINFO_OPMODE_ADHOC &&
+			mode != IWINFO_OPMODE_MESHPOINT)
 		return 0;
 
 	sock = nl80211_wpactl_connect(ifname, &local);
@@ -1115,6 +1116,14 @@ static int nl80211_get_ssid(const char *
 		nl80211_hostapd_query(ifname, "ssid", sb.ssid,
 		                      IWINFO_ESSID_MAX_SIZE + 1);
 
+        /* Before trying uci configuration through ubus,
+         * try to get the meshid through wpa_supplciant config.
+         * call to ubus seems to be creating some problems,
+         * especially making the calling process unrepsonsive to Ctrl+C
+         */
+	if (sb.ssid[0] == 0)
+		nl80211_wpactl_query(ifname, "ssid", sb.ssid,
+		                      IWINFO_ESSID_MAX_SIZE + 1);
 	/* failed, try to obtain Mesh ID */
 	if (sb.ssid[0] == 0)
 		iwinfo_ubus_query(res ? res : ifname, "mesh_id",
