--- a/iwinfo_nl80211.c
+++ b/iwinfo_nl80211.c
@@ -391,6 +391,9 @@ static struct nl80211_msg_conveyor * nl8
 	while (err > 0)
 		nl_recvmsgs(nls->nl_sock, cv->cb);
 
+	if (err)
+		return NULL;
+
 	return &rcv;
 
 err:
@@ -2101,16 +2104,21 @@ static int nl80211_get_scanlist_cb(struc
 
 static int nl80211_get_scanlist_nl(const char *ifname, char *buf, int *len)
 {
-	struct nl80211_msg_conveyor *req;
+	struct nl80211_msg_conveyor *req, *scan_res = NULL;;
 	struct nl80211_scanlist sl = { .e = (struct iwinfo_scanlist_entry *)buf };
 
 	req = nl80211_msg(ifname, NL80211_CMD_TRIGGER_SCAN, 0);
 	if (req)
 	{
-		nl80211_send(req, NULL, NULL);
+		scan_res = nl80211_send(req, NULL, NULL);
 		nl80211_free(req);
 	}
 
+	if (!scan_res)
+	{
+		return -1;
+	}
+
 	nl80211_wait("nl80211", "scan", NL80211_CMD_NEW_SCAN_RESULTS);
 
 	req = nl80211_msg(ifname, NL80211_CMD_GET_SCAN, NLM_F_DUMP);
